pkg/api/v1/conversion.go

```diff
@@ -199,6 +208,104 @@ func addConversionFuncs(scheme *runtime.Scheme) {
+		return err
+	}
+	if err := Convert_v1_ObjectMeta_To_api_ObjectMeta(&in.ObjectMeta, &out.ObjectMeta, s); err != nil {
+		return err
+	}
+	if err := Convert_v1_ReplicationControllerSpec_to_extensions_ReplicaSetSpec(&in.Spec, &out.Spec, s); err != nil {
+		return err
+	}
+	if err := Convert_v1_ReplicationControllerStatus_to_extensions_ReplicaSetStatus(&in.Status, &out.Status, s); err != nil {
+		return err
+	}
+	return nil
+}
+
+func Convert_v1_ReplicationControllerSpec_to_extensions_ReplicaSetSpec(in *ReplicationControllerSpec, out *extensions.ReplicaSetSpec, s conversion.Scope) error {
+	if defaulting, found := s.DefaultingInterface(reflect.TypeOf(*in)); found {
+		defaulting.(func(*ReplicationControllerSpec))(in)
+	}
+	out.Replicas = int(*in.Replicas)
+	if in.Selector != nil {
+		out.Selector = new(unversioned.LabelSelector)
+		for labelKey, labelValue := range in.Selector {
+			labels.AddLabelToSelector(out.Selector, labelKey, labelValue)
+		}
+	}
+	if in.Template != nil {
+		if err := Convert_v1_PodTemplateSpec_To_api_PodTemplateSpec(in.Template, &out.Template, s); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func Convert_v1_ReplicationControllerStatus_to_extensions_ReplicaSetStatus(in *ReplicationControllerStatus, out *extensions.ReplicaSetStatus, s conversion.Scope) error {
+	if defaulting, found := s.DefaultingInterface(reflect.TypeOf(*in)); found {
+		defaulting.(func(*ReplicationControllerStatus))(in)
+	}
+	out.Replicas = int(in.Replicas)
+	out.FullyLabeledReplicas = int(in.FullyLabeledReplicas)
+	out.ObservedGeneration = in.ObservedGeneration
+	return nil
+}
+
+func Convert_extensions_ReplicaSet_to_v1_ReplicationController(in *extensions.ReplicaSet, out *ReplicationController, s conversion.Scope) error {
+	if defaulting, found := s.DefaultingInterface(reflect.TypeOf(*in)); found {
+		defaulting.(func(*extensions.ReplicaSet))(in)
+	}
+	if err := api.Convert_unversioned_TypeMeta_To_unversioned_TypeMeta(&in.TypeMeta, &out.TypeMeta, s); err != nil {
+		return err
+	}
+	if err := Convert_api_ObjectMeta_To_v1_ObjectMeta(&in.ObjectMeta, &out.ObjectMeta, s); err != nil {
+		return err
+	}
+	if err := Convert_extensions_ReplicaSetSpec_to_v1_ReplicationControllerSpec(&in.Spec, &out.Spec, s); err != nil {
+		return err
+	}
+	if err := Convert_extensions_ReplicaSetStatus_to_v1_ReplicationControllerStatus(&in.Status, &out.Status, s); err != nil {
+		return err
+	}
+	return nil
+}
+
+func Convert_extensions_ReplicaSetSpec_to_v1_ReplicationControllerSpec(in *extensions.ReplicaSetSpec, out *ReplicationControllerSpec, s conversion.Scope) error {
+	if defaulting, found := s.DefaultingInterface(reflect.TypeOf(*in)); found {
+		defaulting.(func(*extensions.ReplicaSetSpec))(in)
+	}
+	out.Replicas = new(int32)
+	*out.Replicas = int32(in.Replicas)
+	if in.Selector != nil {
+		selector, err := unversioned.LabelSelectorAsMap(in.Selector)
+		if err != nil {
+			// We should never hit this error since we always come from RCs
```

Comment:
What enforces that you can only round-trip from RC->RS->RC and not just
RS->RC (which is lossy)?

On Tue, Apr 26, 2016 at 12:53 AM, Michail Kargakis <notifications@github.com
> wrote:

> In pkg/api/v1/conversion.go
> <https://github.com/kubernetes/kubernetes/pull/24733#discussion_r61041959>
> :
>
> > +	if err := Convert_extensions_ReplicaSetStatus_to_v1_ReplicationControllerStatus(&in.Status, &out.Status, s); err != nil {
> > +		return err
> > +	}
> > +	return nil
> > +}
> > +
> > +func Convert_extensions_ReplicaSetSpec_to_v1_ReplicationControllerSpec(in *extensions.ReplicaSetSpec, out *ReplicationControllerSpec, s conversion.Scope) error {
> > +	if defaulting, found := s.DefaultingInterface(reflect.TypeOf(*in)); found {
> > +		defaulting.(func(*extensions.ReplicaSetSpec))(in)
> > +	}
> > +	out.Replicas = new(int32)
> > +	*out.Replicas = int32(in.Replicas)
> > +	if in.Selector != nil {
> > +		selector, err := unversioned.LabelSelectorAsMap(in.Selector)
> > +		if err != nil {
> > +			// We should never hit this error since we always come from RCs
>
> RCs don't use the new label selector and based on this conversion
> (rc->rs->rc) we should always get back the same RC we handed over.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24733/files/adcc6ba7decff78b5a2ddf137f8cec1f88c12b90#r61041959>
>